### PR TITLE
SONAR-6644 fix debt when characteristic is none

### PR DESCRIPTION
**Pull request by simonbrandhof**
SONAR-6644 fix debt when characteristic is none

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/rule/RuleOperations.java [hunk 1]
```diff
@@ -111,7 +111,7 @@ public boolean updateRule(RuleDto ruleDto, @Nullable CharacteristicDto newSubCha
       // No sub-characteristic is given -> disable rule debt if not already disabled
     } else {
       // Rule characteristic is not already disabled -> update it
-      if (ruleDto.getSubCharacteristicId() == null || !RuleDto.DISABLED_CHARACTERISTIC_ID.equals(ruleDto.getSubCharacteristicId())) {
+      if (ruleDto.getSubCharacteristicId() == null || RuleDto.DISABLED_CHARACTERISTIC_ID != ruleDto.getSubCharacteristicId()) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RuleOperations.java]: MAJOR Change this comparison to use the equals method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: server/sonar-server/src/main/java/org/sonar/server/debt/DebtModelBackup.java [hunk 2]
```diff
@@ -346,9 +346,9 @@ private static CharacteristicDto characteristicByKey(@Nullable final String key,
   @CheckForNull
   private static RuleDebt toRuleDebt(RuleDto rule, DebtModel debtModel) {
     RuleDebt ruleDebt = new RuleDebt().setRuleKey(RuleKey.of(rule.getRepositoryKey(), rule.getRuleKey()));
-    Integer effectiveSubCharacteristicId = rule.getSubCharacteristicId() != null ? rule.getSubCharacteristicId() : rule.getDefaultSubCharacteristicId();
-    DebtCharacteristic subCharacteristic = (effectiveSubCharacteristicId != null && !RuleDto.DISABLED_CHARACTERISTIC_ID.equals(effectiveSubCharacteristicId)) ?
-      debtModel.characteristicById(effectiveSubCharacteristicId) : null;
+    Integer effectiveSubCharacteristicId = rule.getEffectiveSubCharacteristicId();
+    DebtCharacteristic subCharacteristic = (effectiveSubCharacteristicId != null ?
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/debt/DebtModelBackup.java]: MAJOR Remove those useless parentheses. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: sonar-db/src/main/java/org/sonar/db/rule/RuleDto.java [hunk 3]
```diff
@@ -259,6 +259,15 @@ public RuleDto setDefaultSubCharacteristicId(@Nullable Integer defaultSubCharact
   }
 
   @CheckForNull
+  public Integer getEffectiveSubCharacteristicId() {
+    Integer effective = (subCharacteristicId == null ? defaultSubCharacteristicId : subCharacteristicId);
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/rule/RuleDto.java]: MAJOR Remove those useless parentheses. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.

**Comment by julienlancelot** [comment]
LGTM

**Pull request closed without merging**